### PR TITLE
Final polishing and review

### PR DESCRIPTION
**Pull request by F20170604**
I've updated all the notebooks to be used as final project result.

GAN and Autoencoder notebooks are still in a pending PR. I'll recheck them too as soon as they are reviewed.

### CODE REVIEW

**Comment by F20170604** [comment]
@rxwei @BradLarson can I merge this PR?

File: S4TF_Tutorial_1.ipynb [hunk 1]
```diff
@@ -95,7 +95,7 @@
         "\n",
```
Comment (BradLarson) [inline S4TF_Tutorial_1.ipynb:95]: In the introduction above, maybe changing "Welcome to this Colab" to "Welcome to this notebook" would read more cleanly, given that Colab is mentioned later in the sentence.

File: S4TF_Tutorial_1.ipynb [hunk 2]
```diff
@@ -331,7 +331,7 @@
...
         "Python has lists built into the language.\n",
         "However, we will use a library called numpy for this.\n",
-        "Numpy gives you lot's of support functions that are useful when doing Machine Learning.\n",
+        "Numpy gives you lots of support functions that are useful when doing Machine Learning.\n",
```
Comment (BradLarson) [inline S4TF_Tutorial_1.ipynb]: In this, and throughout the notebook, it might be best to capitalize NumPy per the official capitalization of the project: https://www.numpy.org/ . Also "machine learning" probably shouldn't be capitalized here.

File: S4TF_Tutorial_1.ipynb [hunk 3]
```diff
@@ -331,7 +331,7 @@
...
         "\n",
         "Here, you will also see an import statement. This statement makes the entire numpy package available and we can access those symbols using the abbreviated 'np' syntax."
```
Comment (BradLarson) [inline S4TF_Tutorial_1.ipynb]: Maybe you could phrase this like the following:

"In the following, you will see an import statement that makes the entire NumPy package available to Swift. By assigning this imported library to `np`, we can access NumPy symbols using an abbreviated syntax."

File: S4TF_Tutorial_1.ipynb [hunk 4]
```diff
@@ -487,10 +487,10 @@
           "output_type": "stream",
           "text": [
             "\r\n",
-            "Use numpy to create a [3,3] dimension array with random number\r\n",
-            "[[0.20239917 0.38537939 0.54653833]\r\n",
-            " [0.84937095 0.49260813 0.59873587]\r\n",
-            " [0.94767786 0.53651341 0.13958872]]\r\n"
+            "Use numpy to create a [3, 3] dimension array with random number\r\n",
```
Comment (BradLarson) [inline S4TF_Tutorial_1.ipynb]:
```suggestion
            "Use NumPy to create a 3 x 3 matrix populated by random numbers\r\n",
```

File: S4TF_Tutorial_1.ipynb [hunk 5]
```diff
@@ -470,14 +470,14 @@
...
-        "outputId": "0eb6c79a-fbfe-4ebd-e3af-4f37105ed95b",
+        "outputId": "dcc3a9f2-9fda-4df6-e80f-ee10397abf58",
         "colab": {
           "base_uri": "https://localhost:8080/",
           "height": 102
         }
       },
       "source": [
-        "print(\"\\nUse numpy to create a [3,3] dimension array with random number\")\n",
+        "print(\"\\nUse numpy to create a [3, 3] dimension array with random number\")\n",
```
Comment (BradLarson) [inline S4TF_Tutorial_1.ipynb]: Are the `\n` prefixes needed in these print() statements?

File: S4TF_Tutorial_1.ipynb [hunk 6]
```diff
@@ -541,11 +541,11 @@
         "* Many of the exercises in the course executes more quickly by using GPU runtime: Runtime | Change runtime type | Hardware accelerator | GPU\n",
```
Comment (BradLarson) [inline S4TF_Tutorial_1.ipynb]: Above this, I'd change

"All usage of Colab in this course is completely free or charge. Even GPU usage is provided free of charge for some hours of usage every day."

to 

"All usage of Colab in this course is completely free of charge. Even GPU usage in Colab is provided free of charge for up to several hours of execution every day."

Comment (BradLarson) [inline S4TF_Tutorial_1.ipynb]:
```suggestion
        "* Many of the exercises in this course execute faster when using the GPU runtime. To enable, select Runtime | Change runtime type | Hardware accelerator | GPU\n",
```

File: S4TF_Tutorial_1.ipynb [hunk 7]
```diff
@@ -541,11 +541,11 @@
...
         "\n",
         "**Some final words on Colab**\n",
-        "*   You execute each cell in order, you can edit & re-execute cells if you want\n",
+        "*   You execute each cell in order, you can edit & re-execute cells if you want.\n",
         "*   Sometimes, this could have unintended consequences. For example, if you add a dimension to an array and execute the cell multiple times, then the cells after may not work. If you encounter problem reset your environment:\n",
```
Comment (BradLarson) [inline S4TF_Tutorial_1.ipynb]:
```suggestion
        "*   Sometimes, this could have unintended consequences. For example, if you add a dimension to an array and execute the cell multiple times, the cells that follow may not work. If you encounter problems, reset your environment:\n",
```

File: S4TF_Tutorial_2.ipynb [hunk 8]
```diff
@@ -144,7 +144,7 @@
       "metadata": {
         "id": "Q_XrcNGNbejX",
         "colab_type": "code",
-        "outputId": "e38390ed-c137-4490-9e1b-b45a4c1763f2",
+        "outputId": "e9270b8b-c671-4b1a-c8f7-d960b0f9f7e6",
         "colab": {
           "base_uri": "https://localhost:8080/",
           "height": 136
```
Comment (BradLarson) [inline S4TF_Tutorial_2.ipynb:150]: Near the top, for "Welcome to this Colab where you will train your first Machine Learning model!", I'd not capitalize Machine Learning here (same with anywhere else in this notebook or subsequent notebooks).

File: S4TF_Tutorial_2.ipynb [hunk 9]
```diff
@@ -210,7 +210,7 @@
         "\n",
```
Comment (BradLarson) [inline S4TF_Tutorial_2.ipynb:210]: Above this, for "Of course, it would be simple enough to create a conventional Python function", should that be "Swift function"?

File: S4TF_Tutorial_2.ipynb [hunk 10]
```diff
@@ -210,7 +210,7 @@
...
         "### Select the model\n",
```
Comment (BradLarson) [inline S4TF_Tutorial_2.ipynb:211]: Above this, I'd replace "For the Celsius to Fahrenheit Conersion problem, the model defines the relationship between the Celsius Temperature and Fahrenheit Temperatures."

with

"For the Celsius to Fahrenheit conversion problem, the model defines the relationship between the Celsius temperatures and Fahrenheit temperatures."

File: S4TF_Tutorial_2.ipynb [hunk 11]
```diff
@@ -210,7 +210,7 @@
...
         "\n",
-        "We need to select the kind of model to train. We will use simplest possible model we can, a Dense network. Since the problem is straightforward, this network will require only a single layer, with a single neuron. \n",
+        "We need to select the kind of model to train. We will use simplest possible model we can, a Dense network. Since the problem is straightforward, this network will require only a single layer, with a single neuron.\n",
         "\n"
```
Comment (BradLarson) [inline S4TF_Tutorial_2.ipynb:214]: Below this, "Let's define an CelsiusToFahrenheit" should be "Let's define a CelsiusToFahrenheit struct" 

Also below that "The single value is a Tensor matrix" should just be "The single value is a Tensor"

and below that "The single value is a tensor matrix" should be "The single value is a tensor"

and below that, in the code, "l0" isn't the most descriptive name. Would a better choice be "neuron", to show that it's a dense layer with a single neuron?

File: S4TF_Tutorial_2.ipynb [hunk 12]
```diff
@@ -275,7 +275,7 @@
         "  </td></tr>\n",
         "</table>\n",
         "\n",
-        "Swift for TensorFlow has many [optimization algorithms](https://github.com/rxwei/DeepLearning/blob/master/Sources/DeepLearning/Optimizer.swift) available for training. This model uses the Adam optimizer that implements the *[Adam Algorithm](http://arxiv.org/abs/1412.6980). \n",
+        "Swift for TensorFlow has many [optimization algorithms](https://github.com/rxwei/DeepLearning/blob/master/Sources/DeepLearning/Optimizer.swift) available for training. This model uses the Adam optimizer that implements the *[Adam Algorithm](http://arxiv.org/abs/1412.6980)*. \n",
```
Comment (BradLarson) [inline S4TF_Tutorial_2.ipynb]:
```suggestion
        "Swift for TensorFlow has many [optimization algorithms](https://github.com/tensorflow/swift-apis/tree/master/Sources/TensorFlow/Optimizers) available for training. This model uses the Adam optimizer that implements the *[Adam Algorithm](http://arxiv.org/abs/1412.6980)*. \n",
```

File: S4TF_Tutorial_2.ipynb [hunk 13]
```diff
@@ -275,7 +275,7 @@
...
         "\n",
         "During training, the optimizer function is used to calculate adjustments to the model's internal variables. The goal is to adjust the internal variables until the model (which is really a math function) mirrors the actual equation for converting Celsius to Fahrenheit.\n",
         "\n",
```
Comment (BradLarson) [inline S4TF_Tutorial_2.ipynb:281]: Below this, "optimizer" should probably not be capitalized. The sentence "Too large, and accuracy goes down." could maybe be improved with "Too large, and accuracy may decrease or training could diverge entirely."

File: S4TF_Tutorial_2.ipynb [hunk 14]
```diff
@@ -396,31 +396,31 @@
         {
```
Comment (BradLarson) [inline S4TF_Tutorial_2.ipynb:392]: Above this in the code, I'm not sure "truths" should be plural. "truth" might be a better name for that parameter.

Also, you have an accuracy function, but it's never used. I believe you intended to track that and then plot the accuracy results below. The text refers to visualizing accuracy, but it's not shown on the plot.

File: S4TF_Tutorial_2.ipynb [hunk 15]
```diff
@@ -396,31 +396,31 @@
...
+            "Epoch 380: Loss: 46.829556\n",
+            "Epoch 400: Loss: 37.66132\n",
+            "Epoch 420: Loss: 30.099195\n",
+            "Epoch 440: Loss: 23.904799\n",
+            "Epoch 460: Loss: 18.865921\n",
+            "Epoch 480: Loss: 14.795743\n",
+            "Epoch 500: Loss: 11.531126\n"
           ],
           "name": "stdout"
         }
```
Comment (BradLarson) [inline S4TF_Tutorial_2.ipynb:422]: Just below this, I think the emphasis should be on _see_, not _more_.

File: S4TF_Tutorial_2.ipynb [hunk 16]
```diff
@@ -497,7 +497,7 @@
       "source": [
         "## Use the trained model to make predictions\n",
         "\n",
-        "We've trained a model and demonstrated that it's good—but not perfect—at conerting temperatures. Now let's use the trained model to make some predictions on [unlabeled examples](https://developers.google.com/machine-learning/glossary/#unlabeled_example); that is, on examples that contain features but not a label.\n",
+        "We've trained a model and demonstrated that it's good—but not perfect—at converting temperatures. Now let's use the trained model to make some predictions on [unlabeled examples](https://developers.google.com/machine-learning/glossary/#unlabeled_example); that is, on examples that contain features but not a label.\n",
         "\n",
         "In real-life, the unlabeled examples could come from lots of different sources including apps, CSV files, and data feeds. For now, we're going to manually provide degree Celsius Temperatures to predict their respective degree Fahrenheit Temperatures."
```
Comment (BradLarson) [inline S4TF_Tutorial_2.ipynb]:
```suggestion
        "In real-life, the unlabeled examples could come from lots of different sources including apps, CSV files, and data feeds. For now, we're going to manually provide degree Celsius temperatures to predict their respective degree Fahrenheit temperatures."
```

File: S4TF_Tutorial_3.ipynb [hunk 17]
```diff
@@ -406,7 +406,7 @@
         "## Reading the data\n",
         "\n",
         "We need to read the file into an array of bytes which can further be passed to `readMNIST` function to convert it into a tensor array. \\\n",
```
Comment (BradLarson) [inline S4TF_Tutorial_3.ipynb]:
```suggestion
        "We need to read the file into an array of bytes which can further be passed into the `readMNIST` function to convert it into a tensor array. \n",
```

File: S4TF_Tutorial_3.ipynb [hunk 18]
```diff
@@ -406,7 +406,7 @@
...
-        "The value of each pixel in the image data is an integer in the range `[0,255]`. For the model to work properly, these values need to be normalized to the range `[0,1]`. So here we normalize the image by dividing the pixel value by 255, and then apply it to each image in the test and train datasets. \\\n",
+        "The value of each pixel in the image data is an integer in the range `[0, 255]`. For the model to work properly, these values need to be normalized to the range `[0, 1]`. So here we normalize the image by dividing the pixel value by 255, and then apply it to each image in the test and train datasets. \\\n",
```
Comment (BradLarson) [inline S4TF_Tutorial_3.ipynb]: The line breaks at the end of each of these lines got messed up somehow, turning them into slashes. All of the lines in this paragraph should have the linebreaks corrected at the end of them.

File: S4TF_Tutorial_3.ipynb [hunk 19]
```diff
@@ -406,7 +406,7 @@
...
         "The tensor array we create is in the form of NHWC format which is TensorFlow's default. \\\n",
         "N refers to the number of images in a batch. \\\n",
         "H refers to the number of pixels in the vertical (height) dimension. \\\n",
```
Comment (BradLarson) [inline S4TF_Tutorial_3.ipynb]: Below this, "grabbing minibatches of Images" should have images be lowercased.

Also, "Use Adam Algorithm to update the model's variables." should be "Use the Adam optimizer to update the model's variables."

File: S4TF_Tutorial_3.ipynb [hunk 20]
```diff
@@ -1764,7 +1763,7 @@
         "Experiment with different models and see how the accuracy results differ. In particular change the following parameters:\n",
```
Comment (BradLarson) [inline S4TF_Tutorial_3.ipynb:1767]: Above this, can the plot warning be fixed, so that it doesn't show in the notebook during execution?

File: S4TF_Tutorial_3.ipynb [hunk 21]
```diff
@@ -1764,7 +1763,7 @@
...
         "*   Set training epochs set to 1\n",
         "*   Number of neurons in the Dense layer following the Flatten one. For example, go really low (e.g. 10) in ranges up to 512 and see how accuracy changes\n",
```
Comment (BradLarson) [inline S4TF_Tutorial_3.ipynb]:
```suggestion
        "*   Adjust the number of neurons in the first Dense layer. For example, go really low (e.g. 10) in ranges up to 512 and see how accuracy changes.\n",
```

File: S4TF_Tutorial_3.ipynb [hunk 22]
```diff
@@ -1764,7 +1763,7 @@
         "Experiment with different models and see how the accuracy results differ. In particular change the following parameters:\n",
         "*   Set training epochs set to 1\n",
```
Comment (BradLarson) [inline S4TF_Tutorial_3.ipynb]:
```suggestion
        "*   Set training epochs set to 1.\n",
```

File: S4TF_Tutorial_3.ipynb [hunk 23]
```diff
@@ -1764,7 +1763,7 @@
...
         "*   Number of neurons in the Dense layer following the Flatten one. For example, go really low (e.g. 10) in ranges up to 512 and see how accuracy changes\n",
-        "*   Add additional Dense layers between the Flatten and the final Dense(10,  activation = Softmax), experiment with different units in these layers\n",
+        "*   Add additional Dense layers between the Flatten and the final Dense(10, activation = Softmax), experiment with different units in these layers\n",
```
Comment (BradLarson) [inline S4TF_Tutorial_3.ipynb]:
```suggestion
        "*   Add additional Dense layers between the Flatten layer and the final Dense layer, experiment with different units in these layers.\n",
```

File: S4TF_Tutorial_3.ipynb [hunk 24]
```diff
@@ -1764,7 +1763,7 @@
...
         "*   Don't normalize the pixel values, and see the effect that has\n",
```
Comment (BradLarson) [inline S4TF_Tutorial_3.ipynb]:
```suggestion
        "*   Don't normalize the pixel values, and see the effect that has.\n",
```

File: S4TF_Tutorial_3.ipynb [hunk 25]
```diff
@@ -1764,7 +1763,7 @@
...
         "\n",
         "\n",
```
Comment (BradLarson) [inline S4TF_Tutorial_3.ipynb:1773]: Below this: "Remember to enable GPU" should maybe be "Remember to enable GPU acceleration"

File: S4TF_Tutorial_4.ipynb [hunk 26]
```diff
@@ -37,7 +37,7 @@
       "metadata": {
```
Comment (BradLarson) [inline S4TF_Tutorial_4.ipynb:37]: For this notebook, it looks like many of the comments I made on notebook 3 can be brought across, too.

**Comment by BradLarson** [comment]
For Tutorial 5:

"We'll build an image classifier using Layer and load data by creating Training and Validation Tensors of Images as well as their corresponding Labels" -> "We'll build an image classifier using Layer and load data by creating training and validation tensors of images as well as their corresponding labels."

"around the following concepts" -> "around the following concepts:"

"a filtered version of Dogs vs. Cats" -> "a filtered version of the Dogs vs. Cats"

For the `// Dataset download.` comment, I'd remove the comment and instead make a title or sentence before this code block saying what this is.

"has the following directory structure." -> "has the following directory structure:"

I'd remove the `// List of paths created.` comment.

"Decode contents of these images and convert it into proper grid format as per their RGB content." -> "Decode the contents of these images into their RGB bytes."

"as neural networks prefer to deal with small input values." -> "to better match the range expected by the initial neural network weights."

"The tensor function takes image path as input and outputs a tuple of resized image tensor and it’s corresponding label." -> "The `tensor(path:)` function takes the image path as input and outputs a tuple of the resized image tensor and its corresponding label."

"The tensors function" -> "The tensors(fromList:, valueCount:) function"

"it gives number of images done as print statements" -> "it prints the number of images processed"

Remove the // Tensor Dataset creation. comment.

"We can visualize our training images by creating functions to plot images through" -> "We can visualize our training images by creating functions to plot images from" (We should not be italicized)

```
func plotImages(fromPath path: String) {
```

should be 

```
func plotImages(from path: String) {
```

Remove // A classifier. comment.

"We would also like to look at training" -> "We would also like to look at the training"

"It's time we train our network." -> "It's time to train our network:"

**Comment by BradLarson** [comment]
I'll review notebooks 6, 7, and 8 tomorrow, because I'm unfortunately out of time for today.

If these last three are blockers, you potentially could split out each notebook into a separate pull request so it could be approved individually. That doesn't need to be done right now, but if things drag on you could look into doing that.

**Comment by Ayush517** [comment]
I'm merging these notebooks and opening up 3 new PRs for Tutorials 6, 7 and 8.

**Comment by BradLarson** [comment]
For Tutorial 6:

"We'll now download and unzip the dataset using subprocess library via Python Interoperability." -> "We'll now download and unzip the dataset using the `subprocess` library via Python interoperability."

The line "We can list the directories" looks like it was accidentally code-formatted, when it should be normal text.

"Let's look at how many cats and dogs images we have in our training and validation directory" -> "Let's look at how many cats and dogs images we have in our training and validation directory:"


"We can visualize" -> "We" shouldn't be italicized. Also, this shouldn't be a bulleted list item.

"In Swift for TensorFlow we can implement this using the different Python Libraries through Python Interoperability." -> "In Swift for TensorFlow we can implement this using the different Python libraries through Python interoperability."

"We can simply pass different transformations we would want to our dataset as a form of arguments and it will take care of applying it to the dataset during our training process." -> "We can simply apply different transformations we would want to our dataset images and they will be applied during our training process."

"We perform following Image Augmentation techniques randomly on an image:" -> "We perform the following image augmentation techniques randomly on an image:"

"Adding Noise" -> "Adding noise"

"Apply or not apply Image Augmentation Techniques, based on type of data." -> "Apply or don't apply image augmentation techniques, based on the type of data."

"The resizedImage(fromPath:augmented:) function takes 2 inputs are:" -> "The resizedImage(fromPath:augmented:) function takes 2 inputs:"

The function comments in the following code block ("// Function `resizedImage(" and the one before the other function) can be removed.

The // Test Dataset creation. comment can be removed.

"Since we are output a softmax categorization" -> "Since we output a softmax categorization"

Is there a conclusion section needed after the final plots?

**Comment by BradLarson** [comment]
For notebook 7, there are TODOs around image augmentation. Could the same code from notebook 6 be used for this?

I think some of my same wording comments for notebook 6 can be applied to 7.

**Comment by F20170604** [comment]
> For notebook 7, there are TODOs around image augmentation. Could the same code from notebook 6 be used for this?

Yes, the same code could be used from notebook 6. But I wanted to give independence to the student to write his own code if he wants to use some other libraries for image augmentation. That's why I left it.

> I think some of my same wording comments for notebook 6 can be applied to 7.

Yes, they can be applied on both tutorials 7 and 8 as well

**Comment by BradLarson** [comment]
For tutorial 8:

at the "for filename in filelist {" code block, that's a really long list of filenames. Can you shorten this to only print 10 or so of them?

It looks like many of the same comments I had for Tutorial 6 for wording also apply to tutorial 8.

**Comment by F20170604** [comment]
@Brad done all of them.

**Comment by F20170604** [comment]
Whenever we bring all these notebooks to core documentation, there will still be URL changes in all notebooks regarding Colab and GitHub links as well as the download links of datasets inside the notebooks.

**Review by BradLarson — APPROVED** [review]
Why don't you pull these in as they are. We can still edit them later, but let's get this in for the final review of the project.

I forget, do you have a core writeup document describing the project, what was accomplished, and what was learned, or will you edit the README to host that information?

**Comment by F20170604** [comment]
> Why don't you pull these in as they are. We can still edit them later, but let's get this in for the final review of the project.

Cool 👍 

> I forget, do you have a core writeup document describing the project, what was accomplished, and what was learned, or will you edit the README to host that information?

No, not yet. I am hoping to write a blog today before the deadline but my classes are till 8 pm. So, if I don't get the time, I'll edit the README to host that info.

**Comment by F20170604** [comment]
It's here!

https://medium.com/gsoc-19/the-end-31903840b090
